### PR TITLE
Avoid reading the genesis json file twice upon startup

### PR DESCRIPTION
**Pull request by nikurt**
`nearcore` reads the genesis json file once to get GenesisConfig, and the second time to read the records and get Genesis.
Also, confusingly, the function `load_config_without_genesis_records` does read the genesis records in all cases.

This PR drops `load_config_without_genesis_records` in favor of `load_config` and ensures the genesis json file is read exactly once.

Additionally, this PR adds a command line option to skip genesis validation.

Next steps: Figure out why opening RocksDB takes minutes.

Test results on an archival testnet node:
`neard-1.23 run`: 115 seconds from start until starting to open RocksDB
`neard-1.23-with-fix run`: 90 seconds from start until starting to open RocksDB
`neard-1.23-with-fix --unsafe-fast-startup run`: 67 seconds from start until starting to open RocksDB

Issue #4791

### CODE REVIEW

File: core/chain-configs/src/genesis_config.rs [hunk 1]
```diff
@@ -468,37 +468,75 @@ impl GenesisJsonHasher {
...
         let genesis = Self { config, records: GenesisRecords(vec![]), records_file };
-        validate_genesis(&genesis);
+        match genesis_validation {
+            GenesisValidationMode::Full => {
+                validate_genesis(&genesis);
+            }
+            GenesisValidationMode::UnsafeFast => {
+                warn!("Skipped genesis validation");
+            }
+        }
```
Comment (matklad) [inline core/chain-configs/src/genesis_config.rs:507]: Might make sense to dedupe the code. Usually it's best when there's one "primary constructor", everything else delegates to.

File: core/chain-configs/src/genesis_config.rs [hunk 2]
```diff
@@ -468,37 +468,75 @@ impl GenesisJsonHasher {
...
             serde_json::from_reader(reader).expect("Failed to deserialize the genesis records.");
-        validate_genesis(&genesis);
+        match genesis_validation {
+            GenesisValidationMode::Full => {
+                validate_genesis(&genesis);
+            }
+            GenesisValidationMode::UnsafeFast => {
+                warn!("Skipped genesis validation");
+            }
+        }
```
Comment (matklad) [inline core/chain-configs/src/genesis_config.rs:523]: Oh, with thee copies, I feel strongly that we should dedupe this

File: core/chain-configs/src/genesis_config.rs [hunk 3]
```diff
@@ -468,37 +468,75 @@ impl GenesisJsonHasher {
...
+    Full,
+    UnsafeFast,
+}
+
 impl Genesis {
-    pub fn new(config: GenesisConfig, records: GenesisRecords) -> Self {
+    pub fn new(
+        config: GenesisConfig,
+        records: GenesisRecords,
+        genesis_validation: GenesisValidationMode,
```
Comment (matklad) [inline core/chain-configs/src/genesis_config.rs:480]: All calls to `new` pass in a literal `GenesisValidationMode` so perhaps this signature doesn't need to change?

File: runtime/runtime-params-estimator/src/main.rs [hunk 4]
```diff
@@ -117,7 +118,7 @@ fn main() -> anyhow::Result<()> {
             )
             .expect("failed to init config");
 
-            let near_config = load_config(&state_dump_path);
+            let near_config = load_config(&state_dump_path, GenesisValidationMode::Full);
```
Comment (bowenwang1996) [inline runtime/runtime-params-estimator/src/main.rs:121]: Do we really need to validate genesis in parameter estimator? @matklad

**Review by bowenwang1996 — COMMENTED** [review]
@nikurt could you document somewhere how to start neard without genesis validation?

File: core/chain-configs/src/genesis_config.rs [hunk 5]
```diff
@@ -470,39 +470,77 @@ impl GenesisJsonHasher {
...
+        genesis.validate(genesis_validation)
     }
 
+    fn validate(self, genesis_validation: GenesisValidationMode) -> Self {
+        match genesis_validation {
+            GenesisValidationMode::Full => {
+                validate_genesis(&self);
+            }
+            GenesisValidationMode::UnsafeFast => {
+                warn!("Skipped genesis validation");
```
Comment (frol) [inline core/chain-configs/src/genesis_config.rs:539]: let's add `target` to the warn!

**Review by frol — APPROVED** [review]

**Comment by nikurt** [comment]
> @nikurt could you document somewhere how to start neard without genesis validation?

@bowenwang1996. Done, added flag documentation.

**Review by bowenwang1996 — APPROVED** [review]

**Review by frol — APPROVED** [review]

File: runtime/runtime-params-estimator/src/main.rs (hunk 4 above)
Comment (matklad) [reply runtime/runtime-params-estimator/src/main.rs:123]: I don't think it matters either way -- I'd expect the performance to be dominated by other stuff, and an extra sanity check won't harm.

**Review by matklad — APPROVED** [review]

**Review by khorolets — APPROVED** [review]

**Pull request merged**